dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/message/ProgramMessageController.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+/** Zubair <rajazubair.asghar@gmail.com> */
+@RestController
+@RequestMapping("/api/messages")
+@RequiredArgsConstructor
+@ApiVersion({DhisApiVersion.DEFAULT, DhisApiVersion.ALL})
+public class ProgramMessageController extends AbstractFullReadOnlyController<ProgramMessage> {
+  private final ProgramMessageService programMessageService;
+  private final ProgramMessageRequestParamsMapper paramsMapper;
+  private final RenderService renderService;
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @GetMapping(produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public ResponseEntity<List<ProgramMessage>> getProgramMessages(
+      ProgramMessageRequestParams requestParams) throws BadRequestException, ConflictException {
+    ProgramMessageQueryParams params = paramsMapper.ValidateAndMap(requestParams);
+
+    return ResponseEntity.ok(programMessageService.getProgramMessages(params));
+  }
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @GetMapping(value = "/scheduled/sent", produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public List<ProgramMessage> getScheduledSentMessage(ProgramMessageRequestParams requestParams)
+      throws BadRequestException, ConflictException {
+    requestParams.setMessageStatus(ProgramMessageStatus.SENT);
+    ProgramMessageQueryParams params = paramsMapper.ValidateAndMap(requestParams);
+
+    return programMessageService.getProgramMessages(params);
+  }
+
+  @RequiresAuthority(anyOf = {F_MOBILE_SENDSMS, F_SEND_EMAIL})
+  @PostMapping(consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public BatchResponseStatus sendMessages(HttpServletRequest request) throws IOException {
+    ProgramMessagesImportBatch batch =
+        renderService.fromJson(request.getInputStream(), ProgramMessagesImportBatch.class);
+
+    for (ProgramMessage programMessage : batch.getProgramMessages()) {
+      programMessageService.validatePayload(programMessage);
+    }
+
+    return programMessageService.sendMessages(batch.getProgramMessages());
+  }
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @PutMapping(value = "/{uid}", produces = APPLICATION_JSON_VALUE)
+  public WebMessage updateProgramMessage(
+      @PathVariable String uid, ProgramMessageUpdateRequest updatedRequest)
+      throws NotFoundException {
+    programMessageService.updateProgramMessage(uid, updatedRequest);
+    return ok(format("ProgramMessage with id %s updated", uid));
+  }
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @DeleteMapping(value = "/{uid}", produces = APPLICATION_JSON_VALUE)
+  public WebMessage deleteProgramMessage(@PathVariable String uid) throws NotFoundException {
```

Comment:
```suggestion
  public WebMessage deleteProgramMessage(@PathVariable UID uid) throws NotFoundException {
```

We could directly use `UID` type so we are sure the format is correct